qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/CassandraMetricsClientLimitMultiplier.java

```diff
@@ -0,0 +1,81 @@
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.qos.ratelimit;
+
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import com.palantir.atlasdb.qos.config.CassandraHealthMetric;
+import com.palantir.atlasdb.qos.config.CassandraHealthMetricMeasurement;
+import com.palantir.atlasdb.qos.config.ImmutableCassandraHealthMetricMeasurement;
+import com.palantir.atlasdb.qos.config.QosCassandraMetricsInstallConfig;
+import com.palantir.atlasdb.qos.config.QosCassandraMetricsRuntimeConfig;
+import com.palantir.atlasdb.qos.config.QosPriority;
+import com.palantir.atlasdb.qos.config.ThrottlingStrategy;
+import com.palantir.cassandra.sidecar.metrics.CassandraMetricsService;
+import com.palantir.remoting3.clients.ClientConfigurations;
+import com.palantir.remoting3.jaxrs.JaxRsClient;
+
+public final class CassandraMetricsClientLimitMultiplier implements ClientLimitMultiplier {
+    private final CassandraMetricsService cassandraMetricClient;
+    private ThrottlingStrategy throttlingStrategy;
+    private Supplier<List<CassandraHealthMetric>> cassandraHealthMetrics;
+
+    private CassandraMetricsClientLimitMultiplier(
+            CassandraMetricsService cassandraMetricsService,
+            ThrottlingStrategy throttlingStrategy,
+            Supplier<List<CassandraHealthMetric>> cassandraHealthMetrics) {
+        this.cassandraMetricClient = cassandraMetricsService;
+        this.throttlingStrategy = throttlingStrategy;
+        this.cassandraHealthMetrics = cassandraHealthMetrics;
+    }
+
+    public static ClientLimitMultiplier create(Supplier<QosCassandraMetricsRuntimeConfig> runtimeConfig,
+            QosCassandraMetricsInstallConfig installConfig) {
+        CassandraMetricsService metricsService = JaxRsClient.create(
+                CassandraMetricsService.class,
+                "qos-service",
+                ClientConfigurations.of(installConfig.cassandraServiceConfig()));
+        ThrottlingStrategy throttlingStrategy = ThrottlingStrategyFactory.getThrottlingStrategy(
+                installConfig.throttlingStrategy());
+        return new CassandraMetricsClientLimitMultiplier(metricsService, throttlingStrategy,
+                () -> runtimeConfig.get().cassandraHealthMetrics());
+    }
+
+    public double getClientLimitMultiplier(QosPriority qosPriority) {
+        if (qosPriority == QosPriority.HIGH) {
+            return 1.0; // don't lower the limit for HIGH priority clients.
+        }
+
+        return throttlingStrategy.getClientLimitMultiplier(getCassandraHealthMetricMeasurements(), qosPriority);
+    }
+
+    private Supplier<List<CassandraHealthMetricMeasurement>> getCassandraHealthMetricMeasurements() {
+        return () -> cassandraHealthMetrics.get().stream().map(metric ->
+                ImmutableCassandraHealthMetricMeasurement.builder()
+                        .currentValue(cassandraMetricClient.getMetric(
```

Comment:
Haven't looked at Cassandra metric client code yet - is this a remote call?

Reply:
Yup, this will be a call to the cassandra-sidecar.

Reply:
Cool, we can consider batching in the future.

Reply:
Hmm... even without batching, a call to get limits could block on fetching the health status of all the cassandra metrics. Maybe this behavior is fine, depending on how the client side QOS limits are consumed, but we might want to instead memoize the last seen health metrics so that reads from QoSService don't block on any further external calls.